plugins/experimental/stale_while_revalidate/stale_while_revalidate.c

```diff
@@ -311,46 +324,65 @@ consume_resource(TSCont cont, TSEvent event ATS_UNUSED, void *edata ATS_UNUSED)
       TSMutexUnlock(state->plugin_config->troot_mutex);
-    } else {
-      TSDebug(PLUGIN_NAME, "In sync path. setting fresh and re-enabling");
-      TSHttpTxnCacheLookupCountGet(state->txn, &lookup_count);
-      if (state->resp_info && ((state->resp_info->status == 500) ||
-                               ((state->resp_info->status >= 502) && (state->resp_info->status <= 504)) || lookup_count > 2)) {
-        TSDebug(PLUGIN_NAME, "Sending stale data as fresh");
-        if (state->plugin_config->log_info.object &&
-            (state->plugin_config->log_info.all || state->plugin_config->log_info.stale_if_error)) {
-          CachedHeaderInfo *chi = get_cached_header_info(state->txn);
-          TSTextLogObjectWrite(state->plugin_config->log_info.object, "stale-if-error: %d - %d < %d + %d %s", (int)state->txn_start,
-                               (int)chi->date, (int)chi->max_age, (int)chi->stale_on_error, state->req_info->effective_url);
-          TSfree(chi);
-        }
-        TSHttpTxnHookAdd(state->txn, TS_HTTP_SEND_RESPONSE_HDR_HOOK, state->main_cont);
-        TSHttpTxnCacheLookupStatusSet(state->txn, TS_CACHE_LOOKUP_HIT_FRESH);
-      } else {
-        TSDebug(PLUGIN_NAME, "Attempting new cache lookup");
-        TSHttpHdrUrlGet(state->req_info->buf, state->req_info->http_hdr_loc, &url_loc);
-        TSHttpTxnNewCacheLookupDo(state->txn, state->req_info->buf, url_loc);
-        TSHandleMLocRelease(state->req_info->buf, state->req_info->http_hdr_loc, url_loc);
-        // TODO add txn translation hook and pass result along, maybe inside continuation?
-        // TSHttpTxnHookAdd(state->txn, TS_HTTP_RESPONSE_TRANSFORM_HOOK, TSTransformCreate(replace_transform, state->txn));
-      }
-      TSHttpTxnReenable(state->txn, TS_EVENT_HTTP_CONTINUE);
+
+      free_state_info(state); // async final free
+      TSDebug(PLUGIN_NAME, "TXN State Freed (async)");
+      return 0;
+      //////////////// RETURN (async done)
     }
-    free_request_info(state->req_info);
-    if (state->resp_info) {
-      free_response_info(state->resp_info);
+
+    // state->async_req == false 
+    //    ---> state->resp_info / state->txn / state->main_cont are valid 
+
+    TSDebug(PLUGIN_NAME, "In sync path. setting fresh and re-enabling");
+
+    switch ( state->resp_info->status ) {
+      case TS_HTTP_STATUS_INTERNAL_SERVER_ERROR:
+      case TS_HTTP_STATUS_BAD_GATEWAY:
+      case TS_HTTP_STATUS_SERVICE_UNAVAILABLE:
+      case TS_HTTP_STATUS_GATEWAY_TIMEOUT:
+
+        TSDebug(PLUGIN_NAME, "Sending stale data as fresh");
+
+         if (state->plugin_config->log_info.object &&
+              (state->plugin_config->log_info.all || state->plugin_config->log_info.stale_if_error)) {
+            CachedHeaderInfo *chi = get_cached_header_info(state->txn);
+            TSTextLogObjectWrite(state->plugin_config->log_info.object, "stale-if-error: %ld - %ld < %ld + %ld %s", state->txn_start,
```

Comment:
same message and same status-values ... but no use of TSHttpTxnNewCacheLookupDo() to "retry".